src/main/java/harmonised/pmmo/util/XP.java

```diff
@@ -194,12 +195,49 @@ public static String getSkillFromTool( String tool )
 		}
 	}
 
-	public static Map<String, Double> getXp( String registryName, JType jType )
+	public static Map<String, Double> getXp( TileEntity tile, JType jType )
 	{
-		return getXp( new ResourceLocation( registryName ), jType );
+		ResourceLocation res = tile.getBlockState().getBlock().getRegistryName();
+		
+		if (TooltipSupplier.tooltipExists( res , jType))
+			return TooltipSupplier.getTooltipData( res , jType, tile );
+		
+		return getXp( res , jType );
+	}
+	
+	public static Map<String, Double> getXp( ItemStack stack, JType jType )
+	{
+		ResourceLocation res = stack.getItem().getRegistryName();
+		
+		if (TooltipSupplier.tooltipExists( res , jType))
+			return TooltipSupplier.getTooltipData( res , jType, stack );
+		
+		return getXp( stack.getItem().getRegistryName() , jType );
+	}
+	
+	public static Map<String, Double> getXp( Entity entity, JType jType )
+	{
+		ResourceLocation res = entity.getType().getRegistryName();
+		
+		if (TooltipSupplier.tooltipExists( res , jType))
+			return TooltipSupplier.getTooltipData( res , jType, entity );
+		
+		return getXp( entity.getType().getRegistryName() , jType );
+	}
+	
+	/**	This method is called to deliberately bypass the public getXP methods
+	 * this should only be used when an action does not require an API check.
+	 * 
+	 * @param registryName the object resourcelocation being passed
+	 * @param jType the JType being passed.
+	 * @return a map of the skill and xp value
+	 */
+	public static Map<String, Double> getXpBypass( ResourceLocation registryName, JType jType )
+	{
+		return getXp( registryName, jType);
 	}
 
-	public static Map<String, Double> getXp( ResourceLocation registryName, JType jType )
+	private static Map<String, Double> getXp( ResourceLocation registryName, JType jType )

```

Comment:
This is probably the biggest change.  To make sure that `getXp` was being called in a way that did not circumnavigate the API checks, 3 new methods of the same name were created to handle the different inputs.  These also correspond to a respective checker in the API, each of which are referencing a unique map based on the input object (ItemStack, TileEntity, Entity). 

As an additional security measure to prevent unintentional calls to `getXp(ResourceLocation, JType)` the method was changed to private.  Since some actions still use that method and do not have reasonable TE, Entity, or Itemstack sources (like plants), a new method `getXPBypass(ResourceLocation, JType)` was created to denote a deliberate call to avoid API checks.